internal/cmd/agent/controller/bundledeployment_controller.go

```diff
@@ -220,6 +231,83 @@ func (r *BundleDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Req
 	return ctrl.Result{}, errutil.NewAggregate(merr)
 }
 
+// copyResourcesFromUpstream copies bd's DownstreamResources, from the downstream cluster's namespace on the management
+// cluster to the destination namespace on the downstream cluster, creating that namespace if needed.
+// If bd does not have any DownstreamResources, this method does not issue any API server calls.
+func (r *BundleDeploymentReconciler) copyResourcesFromUpstream(
+	ctx context.Context,
+	bd *fleetv1.BundleDeployment,
+	logger logr.Logger,
+) error {
+	if !experimental.CopyResourcesDownstreamEnabled() {
+		return nil
+	}
+
+	if len(bd.Spec.Options.DownstreamResources) == 0 {
+		return nil
+	}
+
+	destNS := namespaces.GetDeploymentNS(r.DefaultNamespace, bd.Spec.Options)
+
+	ns := corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: destNS}}
+	if err := r.LocalClient.Get(ctx, types.NamespacedName{Name: ns.Name}, &ns); apierrors.IsNotFound(err) {
+		if err := r.LocalClient.Create(ctx, &ns); err != nil {
+			logger.Info(err.Error())
+			return err
+		}
+
+		logger.V(1).Info("Created namespace to copy resources from upstream", "namespace", ns.Name)
+	}
+
+	for _, rsc := range bd.Spec.Options.DownstreamResources {
+		switch strings.ToLower(rsc.Kind) {
+		case "secret":
+			var s corev1.Secret
+			if err := r.Reader.Get(ctx, client.ObjectKey{Namespace: bd.Namespace, Name: rsc.Name}, &s); err != nil {
+				// The bundle deployment is actually created by the bundle reconciler _before_
+				// these objects are copied to the cluster's namespace, hence retries should happen if
+				// they are not found.
+
+				return err
+			}
+
+			s.Namespace = destNS
+			s.ResourceVersion = ""
+			if s.Labels == nil {
+				s.Labels = map[string]string{}
+			}
+			s.Labels["fleet.cattle.io/bundledeployment"] = bd.Name
+
+			if _, err := controllerutil.CreateOrUpdate(ctx, r.LocalClient, &s, func() error { return nil }); err != nil {
+				return fmt.Errorf("failed to create or update secret %s/%s downstream: %v", bd.Namespace, rsc.Name, err)
+			}
+		case "configmap":
+			var cm corev1.ConfigMap
+			if err := r.Reader.Get(ctx, client.ObjectKey{Namespace: bd.Namespace, Name: rsc.Name}, &cm); err != nil {
+				// The bundle deployment is actually created by the bundle reconciler _before_
+				// these objects are copied to the cluster's namespace, hence retries should happen if
+				// they are not found.
+
+				return err
+			}
+			cm.Namespace = destNS
+			cm.ResourceVersion = ""
+			if cm.Labels == nil {
+				cm.Labels = map[string]string{}
+			}
+			cm.Labels["fleet.cattle.io/bundledeployment"] = bd.Name
+
+			if _, err := controllerutil.CreateOrUpdate(ctx, r.LocalClient, &cm, func() error { return nil }); err != nil {
```

Comment:
Same question here (see the Secret part)

Also...  isn't the code for _Secrets_ and _ConfigMaps_ exactly the same? (aside from the object type)
I wonder if it could be reused.